### DISCUSSION TITLE
creation of tables cannot be rollbacked

### DISCUSSION THREAD

**Discussion by fflorent** [oguimbal/pg-mem#108] [Long term plans]
## Steps to reproduce

Run these instructions:
```
begin;
CREATE TABLE "user1" ("id" SERIAL NOT NULL, "name" text NOT NULL, CONSTRAINT "PK_42" PRIMARY KEY ("id"));
rollback;
select table_name from information_schema.tables where table_name='user1';
```

## Issue

 - In postgresql, the last table is not created, thus it does not appear in table `information_schema.tables`;
 - In pg-mem, the table is created despite the rollback;

Thanks for your work :pray:

**Comment by oguimbal** [comment]
No problem :)
Thanks for the feedback.


Yes, that was an early stage is a choice I made.
Transactions are weakly supported:
- Concurrent transcations are impossible (the second to commit will throw)
- Schema updates are commited immediately

Originally, it was motivated by the fact that pg-mem targets mostly unit/integration tests.
The vast majority of them are operating sequentially, and on a given schema 👉 I didnt feel compeled to implement concurrency & schema transactions.

**I could implement this in the future, but unfortunately, this means a non trivial refactoring. So I would not expect it anytime soon 🤔 Sorry**

**Comment by fflorent** [comment]
No worry! Our initial use case was to ensure that nested transactions (`begin` → `commit` / `rollback`) handled with different clients were well supported by a postgresql client, and a unit test with a postgresql-like client would allow us to document that.

So anyway, we don't miss that much right now, but we found interesting to report that. If that's too hard and there is no plan to support it in the future, you can close the issue, I wouldn't mind :).
